File: Hito_4/Observer/EjemploDetallado/VehicleRoutingOriginal/VehicleRoutingSolucionadoObserver/VehicleRoutingValidator/Handlers/AllLocationsAreVisitedHandler.cs
using VehicleRoutingValidator.VehicleRouting;

namespace VehicleRoutingValidator.Handlers;

public class AllLocationsAreVisitedHandler : BaseHandler
{
    private const string Description = "Some locations are not visited.";
    private const int Depot = VehicleRoutingProblem.Depot;
    private readonly int _numberOfLocations;

    public AllLocationsAreVisitedHandler(int numberOfLocations)
        => _numberOfLocations = numberOfLocations;

    public override bool IsSolutionValid(int[] solution)
    {
        if(AreAllLocationsVisited(solution))
            return base.IsSolutionValid(solution);
        UpdateSubscribers(Description);
        return false;
    }

    private bool AreAllLocationsVisited(int[] solution)
    {
        for(int location = 0; location <= _numberOfLocations; location++)
            if (location != Depot && !solution.Contains(location))
                return false;
        return true;
    }
}

File: Hito_4/Observer/Tests_positivos/Test_Detallado_VehicleRouting/handlers/AllLocationsAreVisitedHandler.py
from ..models import VehicleRoutingProblem
import BaseHandler


class AllLocationsAreVisitedHandler(BaseHandler):
    DESCRIPTION = "Some locations are not visited."
    DEPOT = VehicleRoutingProblem.DEPOT  # Asegúrate de que esto exista como atributo de clase

    def __init__(self, number_of_locations: int):
        super().__init__()
        self._number_of_locations = number_of_locations

    def is_solution_valid(self, solution: list[int]) -> bool:
        if self._are_all_locations_visited(solution):
            return super().is_solution_valid(solution)
        self.update_subscribers(self.DESCRIPTION)
        return False

    def _are_all_locations_visited(self, solution: list[int]) -> bool:
        for location in range(self._number_of_locations + 1):
            if location != self.DEPOT and location not in solution:
                return False
        return True
